src/main/java/com/zaxxer/hikari/pool/PoolBase.java

```diff
@@ -152,38 +152,38 @@ void quietlyCloseConnection(final Connection connection, final String closureRea
    boolean isConnectionAlive(final Connection connection)
    {
       try {
-         try {
-            setNetworkTimeout(connection, validationTimeout);
-
-            final long validationSeconds = (int) Math.max(1000L, validationTimeout) / 1000;
+         setNetworkTimeout(connection, validationTimeout);
 
-            if (isUseJdbc4Validation) {
-               return connection.isValid((int) validationSeconds);
-            }
+         final int validationSeconds = (int) Math.max(1000L, validationTimeout) / 1000;
 
+         boolean valid = true;
+         if (isUseJdbc4Validation) {
+            valid = connection.isValid(validationSeconds);
+         }
+         else {
             try (Statement statement = connection.createStatement()) {
-               if (isNetworkTimeoutSupported != TRUE) {
-                  setQueryTimeout(statement, (int) validationSeconds);
+               if (isNetworkTimeoutSupported == FALSE) {
+                  setQueryTimeout(statement, validationSeconds);
                }
 
                statement.execute(config.getConnectionTestQuery());
-            }
-         }
-         finally {
-            setNetworkTimeout(connection, networkTimeout);
+             }
+          }
 
-            if (isIsolateInternalQueries && !isAutoCommit) {
-               connection.rollback();
-            }
-         }
+          if (valid) {
+             if (isIsolateInternalQueries && !isAutoCommit) {
+                connection.rollback();
+             }
 
-         return true;
+             setNetworkTimeout(connection, networkTimeout);
```

Comment:
This absolutely must be in a ``finally``.

Reply:
not really. if connection is invalid or broken, it gets closed immediately on returning false

Reply:
No.  If the timeout is not reset, the ``close()`` call itself may fail prematurely.

Reply:
before closing this you should have waited for reply or drill down code in closeConnection()

Reply:
Trust me, I did.  The timeout still needs to be reset in isConnectionAlive(), because the close request goes into a queue, and there is no way of knowing how many milliseconds (or seconds) it may sit there before being serviced.

Reply:
ok. to close the dead / broken connection, if the queue takes longer to execute quietlyCloseConnection then connection would be closed anyway in previously set network timeout.

Reply:
for dead / broken connection only call that probably succeeds is close(). all other methods would throw

Reply:
the aim here is to close connection and as per doc, setNetworkTimeout will close connection on timeout. validationTimeout in this case.
attempting to close already broken / invalid connection results in more exceptions here.

Reply:
Just because the query times out does not mean the connection is actually dead (disconnected). It just means that the database did not respond within the user’s configured timeout.  Of course, will still evict the connection.

This change is an unimportant optimization that carries risks.  The existing code has been running without reported problems for quite a long time, which is worth a lot.  

Additionally, the setNetworkTimeout() code paths in drivers like MySQL have already proven to be brittle, and several drivers had only just implemented support (primarily  because of HikariCP) and those code paths are largely untested.

Changing the code buys us very little as far as I can tell, and in fact I find it less readable and more difficult to reason about.